source/agora/test/EnrollmentManager.d

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+
+    Contains tests for the creation of an enrollment data, enrolling as a
+    validator and propagating the information through the network
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.test.EnrollmentManager;
+
+version (unittest):
+
+import agora.common.Amount;
+import agora.common.EnrollmentManager;
+import agora.common.Hash;
+import agora.consensus.data.Block;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import agora.consensus.Genesis;
+import agora.test.Base;
+
+/// test enrollment process
+unittest
+{
+    import std.algorithm;
+    import std.conv;
+    import core.time;
+    import core.thread;
+
+    const NodeCount = 4;
+    const SpendAmount = 40_000 * 10_000_000;
+    auto network = makeTestNetwork(NetworkTopology.Simple, NodeCount);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    assert(network.getDiscoveredNodes().length == NodeCount);
+
+    auto nodes = network.apis.values;
+    auto node_1 = nodes[0];
+    auto node_2 = nodes[1];
+
+    // make transactions which have UTXOs for the node.
+    auto gen_key_pair = getGenesisKeyPair();
+    auto pubkey_1 = node_1.getPublicKey();
+    Transaction[] txs;
+
+    foreach (idx; 0 .. Block.TxsInBlock)
+    {
+        auto input = Input(hashFull(GenesisTransaction), idx.to!uint);
+
+        Transaction tx =
+        {
+            TxType.Freeze,
+            [input],
+            [Output(Amount(SpendAmount), pubkey_1)]
+        };
+
+        auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        txs ~= tx;
+    }
+    txs.each!(tx => node_1.putTransaction(tx));
+    containSameBlocks(nodes, 1).retryFor(8.seconds);
+
+    // get UTXOs
+    Hash[] utxo_hashes;
+    auto utxos = node_1.getUTXOs(pubkey_1, TxType.Freeze, SpendAmount);
+    foreach (key, value; utxos) {
+        utxo_hashes ~= key;
+    }
+
+    // check if a unregisterd UTXO hash exists in the validator set
+    assert(!node_2.hasEnrollment(utxo_hashes[0]));
+
+    // create an Enrollment object to be used for the enrollment process
+    auto enroll = node_1.createEnrollment(utxo_hashes[0]);
+
+    // send a request to enroll as a Validator
+    node_1.enrollValidator(enroll);
+
+    // check if a node contains enrollment data previously sent.
+    size_t times = 0; // Number of times we slept for 50 msecs
+    Start:
+        Thread.sleep(10.msecs);
+        auto has_enroll = node_2.hasEnrollment(utxo_hashes[0]);
+        // Retry if we're too early
+        if (has_enroll == false && times++ < 50)
+            goto Start;
```

Comment:
You don't need a goto here. The code I showed you as an example needed it because there was something after in the loop, but it's not the case here.
E.g.:
```D
while (!node_2.hasEnrollment(...) && times++ < 50)
    Thread.sleep(10.msecs);
assert(times < 50);
```

This will assert if it reaches the 500 msecs timeout

Reply:
Why not just use the `retryFor` function like the rest of the test-suite?

Reply:
![Reasons](https://media.giphy.com/media/x43pXtJShv93a/source.gif)